pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaFlow.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler, StageLogging}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+
+import io.pravega.client.stream.{EventStreamWriter}
+
+import scala.util.control.NonFatal
+import akka.stream.ActorAttributes
+import akka.stream.alpakka.pravega.WriterSettings
+
+@InternalApi private final class PravegaFlowStageLogic[A](val shape: FlowShape[A, A],
+                                                          val scope: String,
+                                                          streamName: String,
+                                                          writerSettings: WriterSettings[A])
+    extends GraphStageLogic(shape)
+    with PravegaWriter
+    with StageLogging {
+
+  private def in = shape.in
+  private def out = shape.out
+
+  def clientConfig = writerSettings.clientConfig
+
+  private var writer: EventStreamWriter[A] = _
+
+  /**
+   * Initialization logic
+   */
+  override def preStart(): Unit =
+    try {
+      writer = createWriter(streamName, writerSettings)
+    } catch {
+      case NonFatal(ex) => failStage(ex)
+    }
+
+  writerSettings.keyExtractor match {
+    case Some(keyExtractor) =>
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          val msg = grab(in)
+          writer.writeEvent(keyExtractor(msg), msg)
```

Comment:
I have a basic query here...

Given that writeEvent returns a future
```
java.util.concurrent.CompletableFuture<java.lang.Void> writeEvent(java.lang.String routingKey, Type event)
```
shouldn't we wait on the future to ensure the event is persisted? Perhaps chain the next actions to this future.

Note: `writer.close()` @ line 73  does invoke `flush()` and ensures all the inflight events are in fact persisted by Pravega.

Reply:
Nice catch I didn't notice it was async ... I will handle this of course, as AFAIU it is needed to insure back pressure

Reply:
Thanks for jumping into the PR @shrids. The docs say the `CompletableFuture` returned from a write will never complete exceptionally. Does that mean it will never timeout either? Can a user define a timeout if they want one?

```
    * @return A completableFuture that will complete when the event has been durably stored on the configured
     *         number of replicas, and is available for readers to see. This future may complete exceptionally
     *         cannot happen, however these exceptions are not transient failures. Failures that occur as a
     *         if this result of connection drops or host death are handled internally with multiple retires and
     *         exponential backoff. So there is no need to attempt to retry in the event of an exception.
```

Reply:
The config specified by `EventWriterConfig` configures the max number of retries and the max backoff time per retry. This config used to create the writer ```io.pravega.client.EventStreamClientFactory#createEventWriter(java.lang.String, io.pravega.client.stream.Serializer<T>, io.pravega.client.stream.EventWriterConfig)```
The EventStreamWriter will retry with the configured number of retries and if all the retries fail the CompletableFuture will be completed exceptionally.

Reply:
Thank you, another question, is the ordering is guaranteed with CompletableFutures, I mean are CompletableFutures completed in the same order they where started ?

Reply:
yes, let's say for given routing key `k` two events `e1` followed by `e2` is written by the writer, and if `cf1` and `cf2` are the futures returned. The ordering guarantee provided by Pravega will ensure that events are persisted in the order `e1` followed by `e2` and the respective futures will be completed once the event is durably persisted.

related doc link : http://pravega.io/docs/latest/pravega-concepts/#ordering-guarantees

Reply:
Cool. We should probably add some notes regarding writing parallelism and ordering guarantees (of Akka Streams and Pravega) to the docs.

Reply:
@cheleb I think this still makes sense to add to the docs ^.